src/node_file.cc

```diff
@@ -2727,6 +2730,229 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+  if (file_url->type != ada::scheme::FILE) {
+    env->isolate()->ThrowException(ERR_INVALID_URL_SCHEME(env->isolate()));
+
+    return false;
+  }
+
+  std::string_view pathname = file_url->get_pathname();
+#if _WIN_32
+  std::string pathname_escaped_slash;
+
+  for (size_t i = 0; (i + 2) < pathname.size(); ++i) {
+    pathname_escaped_slash += pathname[i];
+
+    if (pathname[i] == '\\') pathname_escaped_slash += '\\';
+
+    if (pathname[i] != '%') continue;
+
+    char third = pathname[i + 2] | 0x20;
+
+    bool is_slash = pathname[i + 1] == '2' && third == 102;         // 2f 2F /
+    bool is_forward_slash = pathname[i + 1] == '5' && third == 99;  // 5c 5C \
+
+    if (!is_slash && !is_forward_slash) continue;
+
+    env->isolate()->ThrowException(ERR_INVALID_FILE_URL_PATH(
+        env->isolate(),
+        "File URL path must not include encoded \\ or / characters"));
+
+    return false;
+  }
+
+  std::string decoded_pathname = ada::unicode::percent_decode(
+      std::string_view(pathname_escaped_slash), pathname_escaped_slash.size());
+
+  if (hostname.size() > 0) {
+    // If hostname is set, then we have a UNC path
+    // Pass the hostname through domainToUnicode just in case
+    // it is an IDN using punycode encoding. We do not need to worry
+    // about percent encoding because the URL parser will have
+    // already taken care of that for us. Note that this only
+    // causes IDNs with an appropriate `xn--` prefix to be decoded.
+    *result_file_path =
+        "\\\\" + ada::unicode::to_unicode(hostname) + decoded_pathname;
+
+    return true;
+  }
+
+  char letter = decoded_pathname[1] | 0x20;
+  char sep = decoded_pathname[2];
+
+  // a..z A..Z
+  if (letter < 'a' || letter > 'z' || sep != ':') {
+    env->isolate()->ThrowException(ERR_INVALID_FILE_URL_PATH(
+        env->isolate(), "File URL path must be absolute"));
+
+    return false;
+  }
+
+  *result_file_path = decoded_pathname.substr(1);
+
+  return true;
+#else
+  std::string_view hostname = file_url->get_hostname();
+
+  if (hostname.size() > 0) {
+    std::string error_message =
+        std::string("File URL host must be \"localhost\" or empty on ") +
+        std::string(per_process::metadata.platform);
+    env->isolate()->ThrowException(
+        ERR_INVALID_FILE_URL_HOST(env->isolate(), error_message.c_str()));
+
+    return false;
+  }
+
+  int first_percent = 0;
+
+  for (std::string::size_type i = 0; (i + 2) < pathname.size(); ++i) {
+    if (pathname[i] != '%') continue;
+
+    if (first_percent == 0) {
+      first_percent = i;
+    }
+
+    if (pathname[i + 1] != '2' || (pathname[i + 2] | 0x20) != 102) continue;
+
+    env->isolate()->ThrowException(ERR_INVALID_FILE_URL_PATH(
+        env->isolate(), "File URL path must not include encoded / characters"));
+
+    return false;
+  }
+
+  *result_file_path = ada::unicode::percent_decode(pathname, pathname.size());
+
+  return true;
+#endif
+}
+
+FilePathIsFileReturnType FilePathIsFile(Environment* env,
```

Comment:
This function seems to be the same as https://github.com/nodejs/node/blob/4bb06dbd0a87e5b2430780ddceec08015f8c496a/src/node_file.cc#L1124

maybe we could DRY the implementation of both here? (this not really blocking just a observation)

Reply:
Initially, I was thinking of doing that but I cannot join both functions since I don't know how I can represent the state of the `THROW_IF_INSUFFICIENT_PERMISSIONS`, when the user doesn't have permissions, we should stop the execution of the function, so I need to return not only an `int` but also another state to tell the caller that the user doesn't have permission.

Because of this, I prefer to return an enum for `FilePathIsFile` and not over complicate both implementations.

Reply:
@RafaelGSS might help